content/post/postgres/rfc-extension-packaging-lookup.md

```diff
@@ -0,0 +1,496 @@
+---
+title: "RFC: Extension Packaging & Lookup"
+slug: rfc-extension-packaging-lookup
+date: 2024-11-04T19:07:44Z
+lastMod: 2024-11-04T19:07:44Z
+description: |
+  A proposal to modify the PostgreSQL core so that all files required for an
+  extension live in a directory named for the extension, along with a search
+  path to find extension directories.
+tags: [Postgres, Extensions, RFC, Packaging, Kubernetes, Docker, Packaging, Postgres.app]
+type: post
+draft: true
+---
+
+A few weeks ago, I started [a pgsql-hackers thread] proposing a new extension
+file organization and a search path [GUC] for finding extensions. The
+[discussion] of [Christoph Berg]'s [`extension_destdir` patch][destdir]
+inspired this proposal. These threads cover quite a lot of territory, so I
+thought it would be useful to pull together a more unified, public proposal.
+
+## The Problem
+
+A number of challenges face extension users in various configurations, thanks
+to extension file organization in the Postgres core. The common thread among
+them is the need to add extensions without changing the Postgres installation
+itself.
+
+### Packager Testing
+
+On Debian systems, the user account that creates extension packages lacks
+permission to add files to Postgres install. But testing extensions requires
+installing the extension files where Postgres can find them. Furthermore,
+extensions should ideally be built against a clean Postgres install; adding an
+extension in order to run `make installcheck` would pollute it.
+
+[Christoph's patch][destdir] solves these problems by adding a second lookup
+path for extensions and dynamic modules, so that Postgres can load them
+directly from the package build directory.
+
+Alas, the patch isn't ideal, because it simply specifies a prefix and appends
+the full `pg_config` directory paths to it. For example, if `--sharedir`
+outputs `/opt/share` and `extension_destdir` GUC is set to `/tmp/build/myext`,
+the patch will search in `/tmp/build/myext/opt/share`. This approach works for
+the packaging use case, which explicitly uses full paths with a prefix, but
+would be weird for other use cases.
+
+### Docker Immutability
+
+Docker images are immutable. To install persistent extensions in a running
+Docker container, one must create a persistent volume, map it to
+`SHAREDIR/extensions`, and copy over all the core extensions (or muck with
```

Comment:
Applies here but also elsewhere. I disagree with the distinction of "core extensions", "system extensions" and "user extensions". They are all "extensions" and they should all be treated equally. Not only classifying them introduces unsolicited additional complexity; but also tends to force assumptions that may not be good practices IMO. For example, "core extensions" seem to be required, whereas a production system with tight (read: good) security standards may only want to have the extensions that are going to be used. This is specially true in containerized environments, where you may want to start with an empty list of extensions (well, maybe plpgsql, but that's it) and then only add those that the user wants, whether they are contrib, packages or compiled by hand. I actually believe this should be the case whether you run on containers or not.

Reply:
I'll remove the bit about core extensions here, but I think the question of core/vendor/site distinction is worth getting into. I added it to this proposal while thinking through "Where should extensions be installed by default"? The `pg_config` values are set at compile time, whereas the `extension_path` variable is set at install and/or runtime.

So the question I wanted to address was, "if I run `make install` for an extension, where should it go? I'm not keen on mixing up core and non-core extensions, mainly because, as you guessed, I tend to start with builds that include them all (`make world`. 

I liked the idea of `make install` putting extensions in an appropriate directory regardless of how `extension_path` is set and that keeps core and non-core extensions separate.

But maybe the distinction isn't important? Worth a broader community discussion.

Reply:
Agree on bringing in more PoVs. Here's my more complete reasoning.

> But maybe the distinction isn't important? Worth a broader community discussion.

That's my main point. On one hand, it introduces additional complexity, I think this is objective (more directories, more config flags, more GUCs, you name it).

On the other hand... what value does this distinction bring? I see no added value. Most people if blindly asked to classify into these three groups a given extension would fail for many of them. This distinction seems artificial to me and not linked to any of the following dimensions: extension quality, popularity or usefulness.

BTW what you call "core" extensions I understand are the "contrib" modules that come with Postgres source code. Which

> are not part of the core PostgreSQL system. They are separate mainly because they address a limited audience or are too experimental to be part of the main source tree.
https://www.postgresql.org/docs/current/contrib.html

Calling them "core" bears meaning only with respect to their provenance, but definitely not the other usual assumptions that may come with that word that to my understanding are "fundamental", "same quality level", etc. They are not (despite several being quite useful, of course!) but I see it potentially misleading. A user may want to add them all (because they are core!) when the best security principle is to only bring in those that you will use.

In StackGres, we model all extensions alike. Core/contrib extensions just happen to have the same Git repo as the Postgres project and a `contrib/$extension` subdirectory where the source is located, but that's it.

As to `make install`, I'd be all in that the extension being compiled should go to `pg_config --extdir`, which the user may or may not override with `extension_path` (that's up to the user).

Reply:
> On the other hand... what value does this distinction bring? I see no added value.

Yes, that's the question. Most dynamic programming languages feature the distinction, and a sysadmin may wish to tell them apart just by location. But I'm not too wed to it if the semantics of the eventual solution are nice enough not to need them.

> BTW what you call "core" extensions I understand are the "contrib" modules that come with Postgres source code.

True, except for the [procedural languages that ship with Postgres](https://github.com/postgres/postgres/tree/master/src/pl). So I used `core` to cover "everything that's part of the core project". Could use a different term, tho.

Reply:
My point here is about making things simple: it (almost always) works best for everybody: administrators, users, packagers. Making distinctions about extensions that few people (and I'd argue only highly technical ones) can make does not seem to bring value to end users. If end users cannot tell apart which extensions are "core", which are "packages", which are "installed" or whatever categories we'd establish, I'd not introduce any distinction. And I bet most cannot even tell which ones are core, contrib or otherwise.

Reply:
[Posted to hackers about this](https://postgr.es/m/32400C10-6602-419F-86DB-EB689AC95FC1@justatheory.com).

Reply:
I removed the separate directories in 2c6f9e1.